torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -0,0 +1,208 @@
+#include "torch/csrc/jit/passes/batch_mm.h"
+
+#include "torch/csrc/jit/passes/dead_code_elimination.h"
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/utils/functional.h"
+
+#include <ATen/ATen.h>
+#include <algorithm>
+#include <unordered_map>
+
+namespace torch { namespace jit {
+
+// This pass looks for trees in the graph, where leaves are mm ops, and the inner
+// vertices are add nodes. Once we have such a tree they can be reduced to two
+// concats and a single mm (basically into a single multiply of a wide matrix, with
+// a tall matrix).
+// Such patterns show up mostly in backward of RNNs, since the derivative of many
+// uses of matrix multiplies with same weights forms exactly such a tree
+// (note that it's usually also highly imbalanced i.e. has O(n) depth).
+//
+// This (or any tree of adds of MMs):
+//
+// +------+ +------+   +------+ +------+   +------+
+// |      | |      |   |      | |      |   |      |
+// |  L1  | |  R1  | + |  L2  | |  R2  | = |  O   |
+// |      | |      |   |      | |      |   |      |
+// +------+ +------+   +------+ +------+   +------+
+//
+// can be basically transformed into a single MM which looks like this
+// (we concat all lhs operands, concat rhs operands, do mm):
+//
+//                 +------+
+//                 |      |
+//                 |  R1  |
+//                 |      |
+//                 +------+
+//                 |      |
+//                 |  R2  |
+//                 |      |
+//                 +------+
+// +------+------+ +------+
+// |      |      | |      |
+// |  L1  |  L2  | |  O   |
+// |      |      | |      |
+// +------+------+ +------+
+
+
+
+// Note [Further optimizations]
+// It would be straightforward to extend the TreeToken class to also detect if all
+// MMs had the same lhs/rhs. In such case it's more efficient to expand the lhs
+// and use bmm + sum instead of repeating it in memory via concat.
+
+// Note [Overlapping trees]
+// Additionally it wouldn't be too hard to add support for partially overlapping
+// trees. Right now the it's forbidden in the algorithm (only a single tree will
+// be allowed), so theoretically we might miss some optimization options, especially
+// that the rejected tree could be much larger. I didn't implement that because it's
+// not necessary for the simple RNN cases I saw, so I decided to keep stuff simple.
+// If we ever get around implementing this, the right solution is probably to fuse
+// MMs for the common part, and assume it's an input leaf for the outer two parts
+// (I don't think it's beneficial to recompute, unless the subtree is super small,
+// but let's not get into such details).
+
+// The algorithm we're using is simple. We're iterating through the graph in the
+// topological order and labeling nodes with TreeTokens. Then, we look for roots of
+// the trees we formed and fuse them.
+
+enum class Side {
```

Comment:
nit: move this closer to its use. I kept looking for these tags in the Tree, but it is really just used very locally as an argument to the thing that extracts the batches.